### PR TITLE
Name-keyed posterior assembly via field_order (mechanism)

### PR DESCRIPTION
**Pull request by jhuggins**
Refs #233.

## Problem

`ApproximateDistribution` splits the concatenated flat chain into template fields **positionally** — by `record_template.fields` order and per-field flat sizes, with only a total-flat-size sanity check. Correctness has rested on an unwritten contract: *every caller must concatenate chain columns in `record_template.fields` order.* A backend whose trace orders variables differently (nutpie sorts `posterior.data_vars` alphabetically) silently mislabels draws unless the caller pre-sorts — exactly the bug fixed per-caller in #231.

## Change (mechanism only)

Adds an optional `field_order: list[str] | None` to `ApproximateDistribution.__init__` and `make_posterior`. It names the field each contiguous column-block of the chain occupies. When provided, the flat chain's columns are permuted into `record_template` order (new `_column_permutation` helper) **before** the existing positional split, so columns are aligned to fields by name instead of by position.

- **Default `field_order=None` is unchanged** — every current backend keeps the positional contract byte-for-byte (the reorder is a no-op preprocessing step that only runs when `field_order` is passed).
- Reordering the flat columns up front (rather than building a per-field Record) means all downstream consumers — `draws()`'s `unflatten`, `mean`, the `chains` property — see template-ordered columns and stay correct.

## Validation (when `field_order` is given)

Hardened so misuse fails loudly rather than silently corrupting draws:

- **Requires a template** — `field_order` without `record_template` raises (it names template fields).
- **Permutation-checked for any field count** — a wrong/typo'd name raises, including for single-field templates.
- **Chain width validated before the gather, for any field count** — `c[..., perm]` would otherwise silently drop extra columns (too-wide chain) or clamp out-of-bounds gather indices (too-narrow), then pass the post-gather size check.
- `ApproximateDistribution`'s `Parameters` docstring now documents `record_template` and `field_order` (it's public via mkdocstrings).

## Scope / follow-up

This PR lands the **assembly mechanism + validation** only. Wiring the PyMC inference paths to pass `field_order` (retiring the #231 `keep_names` pre-sort) is **#236**, stacked on this. Branched off `main`.

## Tests (`tests/inference/test_inference.py`)

- `test_field_order_reassembles_by_name` — non-template column order is reassembled by name.
- `test_field_order_none_is_positional` — default path matches the historical layout.
- `test_field_order_must_be_permutation` / `test_field_order_single_field_invalid_permutation_raises` — non-permutation raises (multi- and single-field).
- `test_field_order_chain_too_wide_raises` / `test_field_order_chain_too_narrow_raises` / `test_field_order_single_field_width_mismatch_raises` — width mismatches raise.
- `test_field_order_without_template_raises` — `field_order` without a template raises.

## Test plan

- [x] `pytest tests/inference/test_inference.py` — 95 passed (incl. positional backward-compat)
- [x] `pytest tests/inference/ tests/modeling/` — green
- [ ] CI green

### CODE REVIEW

File: probpipe/inference/_approximate_distribution.py [hunk 1]
```diff
@@ -64,13 +114,29 @@ def __init__(
         weights: ArrayLike | Weights | None = None,
         name: str | None = None,
         record_template: RecordTemplate | None = None,
+        field_order: list[str] | None = None,
     ):
         if not chains:
             raise ValueError("Must provide at least one chain")
 
         self._chains = [jnp.asarray(c) for c in chains]
```
Comment (SoymilkWinsAgain) [inline probpipe/inference/_approximate_distribution.py:136]: The later on code checks if `flat.shape[-1] == total`. If the original chain here has an extra column, `c[..., perm]` will directly select 3 columns and discard the extra column, and subsequent validation will pass. This will silently drop data.

Before reordering, it should be verified that the last dimension of each original chain equals `len(perm) / template total size` before executing the index.

File: probpipe/inference/_approximate_distribution.py [hunk 2]
```diff
@@ -64,13 +114,29 @@ def __init__(
...
         self._concatenated: Array | None = None
 
+        # When the caller's chain columns are laid out in a different
+        # field order than the template — e.g. a backend whose trace
+        # sorts variable names — permute them into ``record_template``
+        # order. The positional split below (and ``draws()`` unflatten)
+        # then map each column to the right field by name rather than by
+        # position, so callers don't have to pre-sort. See issue #233.
+        if (
+            field_order is not None
```
Comment (SoymilkWinsAgain) [inline probpipe/inference/_approximate_distribution.py]: `field_order` is silently ignored when `record_template is None` or when it's a single-field template. Passing this parameter indicates that the caller considers field semantics to exist; a `ValueError` should be thrown if no template is provided. Single-field templates should also validate permutations.

File: tests/inference/test_inference.py [hunk 3]
```diff
@@ -182,6 +182,55 @@ def test_without_template_returns_array(self):
...
+        """A field_order that isn't a permutation of template fields raises."""
+        template = RecordTemplate(a=(), b=())
+        chain = jax.random.normal(jax.random.PRNGKey(0), (5, 2))
+        prior = MultivariateNormal(loc=jnp.zeros(2), cov=jnp.eye(2), name="z")
+        with pytest.raises(ValueError, match="not a permutation"):
+            make_posterior(
+                [chain], parents=(prior,), algorithm="test",
+                record_template=template, field_order=["a", "c"],
+            )
+
```
Comment (SoymilkWinsAgain) [inline tests/inference/test_inference.py:233]: The test is missing a few direct coverages: 
1. field_order + excessively wide chain should raise an error;
2.  excessively narrow chain should give a clear ValueError; 
3. field_order without record_template should raise an error; 
4. invalid field_order for a single field template should raise an error.

**Review by SoymilkWinsAgain — CHANGES_REQUESTED** [review]

File: probpipe/inference/_approximate_distribution.py (hunk 1 above)
Comment (jhuggins) [reply probpipe/inference/_approximate_distribution.py:136]: Fixed in b73f53a. You're right — and it's actually worse for the too-narrow case: JAX gather *clamps* out-of-bounds indices rather than raising, so a too-narrow chain would duplicate the last column silently. Both are masked because the gather always yields exactly `len(perm) == total` columns, so the later size check passes.

Now each chain's last dim is validated against `len(perm)` (the template total) *before* the gather, raising a clear `ValueError` on either too-wide or too-narrow. Covered by `test_field_order_chain_too_wide_raises` and `test_field_order_chain_too_narrow_raises`.

File: probpipe/inference/_approximate_distribution.py (hunk 2 above)
Comment (jhuggins) [reply probpipe/inference/_approximate_distribution.py]: Fixed in b73f53a. The guard is restructured so that whenever `field_order` is given it's validated regardless of field count:

- `field_order` + `record_template is None` → `ValueError` (it names template fields and is meaningless without one).
- The permutation is validated via `_column_permutation` for *any* field count, so a single-field wrong/typo'd name now raises instead of being silently ignored.
- The column reorder + width check still run only for multi-field templates (a single field needs no reorder).

Covered by `test_field_order_without_template_raises` and `test_field_order_single_field_invalid_permutation_raises`.

File: tests/inference/test_inference.py (hunk 3 above)
Comment (jhuggins) [reply tests/inference/test_inference.py:233]: Added all four in b73f53a:
1. `test_field_order_chain_too_wide_raises`
2. `test_field_order_chain_too_narrow_raises`
3. `test_field_order_without_template_raises`
4. `test_field_order_single_field_invalid_permutation_raises`

(The existing `test_field_order_must_be_permutation` already covers the multi-field invalid-permutation case.) Full inference suite: 284 passed, 1 skipped.

File: probpipe/inference/_approximate_distribution.py [hunk 4]
```diff
@@ -64,13 +114,44 @@ def __init__(
...
+            if record_template is None:
+                raise ValueError(
+                    "field_order requires a record_template; it names "
+                    "template fields and is meaningless without one."
+                )
+            # Validates field_order is a permutation of the template
+            # fields (raises otherwise) for any field count, so a
+            # single-field typo or wrong name is also caught.
+            perm = _column_permutation(record_template, field_order)
+            if len(record_template.fields) > 1:
```
Comment (SoymilkWinsAgain) [inline probpipe/inference/_approximate_distribution.py:165]: From CodeX:

The width check is still placed inside `if len(record_template.fields) > 1:`. This way, single-field templates only check the field name permutation, not whether the last dimension of the chain equals the template's flat size.

Example: `RecordTemplate(a=(2,)) + chain.shape == (5, 3) + field_order=["a"]` will pass the construction; subsequently, `draws()` will unflatten columns according to the template, consuming only the first two columns, and extra columns may be silently discarded. This is the same type of data integrity issue as the recently fixed multi-field issue where "extra columns are discarded during gather".

Recommendation: If `field_order` is not None and `_column_permutation(...)` succeeds, perform a `c.shape[-1] == len(perm)` check on each chain; the reordering itself can continue to be performed only in multi-field cases.

File: probpipe/inference/_approximate_distribution.py
Comment (SoymilkWinsAgain) [inline probpipe/inference/_approximate_distribution.py:77]: From CodeX:

`ApproximateDistribution` is a public class, exposed in the documentation via `::: probpipe.ApproximateDistribution`. However, the class documentation only lists `chains/weights/name` in its parameters, omitting the existing `record_template` and the newly added `field_order`. The `make_posterior` documentation already includes `field_order`, and the constructor should also be synchronized, specifically specifying that it is a "contiguous field-block order", not an arbitrary per-column map.

File: tests/inference/test_inference.py [hunk 5]
```diff
@@ -182,6 +182,104 @@ def test_without_template_returns_array(self):
     def test_record_template_property(self, posterior_with_template, template):
         assert posterior_with_template.record_template is template
 
+    def test_field_order_reassembles_by_name(self):
```
Comment (SoymilkWinsAgain) [inline tests/inference/test_inference.py:185]: It is recommended to update the PR description to include the newly added scope.

**Review by SoymilkWinsAgain — CHANGES_REQUESTED** [review]

File: probpipe/inference/_approximate_distribution.py (hunk 4 above)
Comment (jhuggins) [reply probpipe/inference/_approximate_distribution.py:165]: Fixed in 884b80c — good catch (and CodeX). Moved the width check out of the `if len(record_template.fields) > 1:` guard so it runs for any field count; a single-field template with `field_order` now also validates that the chain's last dim equals the template's flat size (`len(perm)`). The reorder itself stays multi-field-only (a single field's permutation is the identity). Added `test_field_order_single_field_width_mismatch_raises`.

File: probpipe/inference/_approximate_distribution.py
Comment (jhuggins) [reply probpipe/inference/_approximate_distribution.py:77]: Fixed in 884b80c. Added `record_template` and `field_order` to the `ApproximateDistribution` class docstring `Parameters` (both were missing; `record_template` predates this PR). They now render under `::: probpipe.ApproximateDistribution`.

File: tests/inference/test_inference.py (hunk 5 above)
Comment (jhuggins) [reply tests/inference/test_inference.py:185]: Will do — updating the PR description now to cover the added validation scope (field_order width check, no-template error, single-field validation).

**Review by SoymilkWinsAgain — APPROVED** [review]
Looks good!
